lib/render/FlowRenderer.cpp

```diff
@@ -297,43 +298,20 @@ FlowRenderer::_paintGL( bool fast )
             /* If the advection is single-directional */
             if( params->GetFlowDirection() == 1 )           // backward integration
                 deltaT *= -1.0f;
-            int numOfSteps = params->GetSteadyNumOfSteps();
-            int total = numOfSteps - (_advection.GetMaxNumOfPart() - 1);
-            int done = 0;
+            long numOfSteps = params->GetSteadyNumOfSteps();
+            //long stepsToGo  = numOfSteps - long(_advection.GetMaxNumOfPart() + 1);
+            //stepsToGo = std::max( stepsToGo, 0l );
             
-            if(_2ndAdvection)
-                total += numOfSteps - (_2ndAdvection->GetMaxNumOfPart() - 1);
-                
-            
-            Progress::Start("Advect particles", total, true);
-            for( size_t i = _advection.GetMaxNumOfPart() - 1;  // existing number of advection steps 
-                 i < numOfSteps && rv == flow::ADVECT_HAPPENED; i++ )
-            {
-                Progress::Update(done++);
-                if (Progress::Cancelled())
-                    return 0;
-                rv = _advection.AdvectOneStep( &_velocityField, deltaT );
-            }
-            if (!_2ndAdvection)
-                Progress::Finish();
+            _advection.AdvectSteps( &_velocityField, deltaT, numOfSteps );
 
             /* If the advection is bi-directional */
             if( _2ndAdvection )
             {
                 assert( deltaT > 0.0f );
                 float   deltaT2 = deltaT * -1.0f;
-                rv = flow::ADVECT_HAPPENED;
-                for( size_t i = _2ndAdvection->GetMaxNumOfPart() - 1; 
-                     i < numOfSteps && rv == flow::ADVECT_HAPPENED; i++ )
-                {
-                    Progress::Update(done++);
-                    if (Progress::Cancelled())
-                        return 0;
-                    rv = _2ndAdvection->AdvectOneStep( &_velocityField, deltaT2 );
-                }
-                Progress::Finish();
```

Comment:
There is a Progress::StartIndefinite method you could use here if you cannot maintain a list of completed jobs.

Reply:
Cool. Does it mean the progress bar stays at 0% for a while, and then jumps to 100% all of a sudden?

Reply:
Yes, but it lets user know vapor has not frozen and you could still use it to allow the user to cancel a flow calculation.

Reply:
@shaomeng were you able to restore the progress meter for flow vis? My recollection was that you disabled it when you were refactoring.

Reply:
yeah, I can add it back but it's gonna show only two states: 0% and 100%, because the FlowRenderer currently won't have any knowledge on how far the advection has gone. Let me bring back the semi-useful progress bar for now.

Reply:
A progress bar is added in 646d0eca .